### ISSUE TITLE
DST Staging Review: Button flickers when clicking an option

### ISSUE THREAD

**Issue by shiragoodman** [department-of-veterans-affairs/vets-design-system-documentation#4390]
_Need help? Please review [how to read a Staging Review ticket](https://depo-platform-documentation.scrollhelp.site/collaboration-cycle/anatomy-of-a-staging-review-issue-ticket). Tag `@platform-governance-team-members` on Slack if you need further assistance._

## Design System Staging Information
**Component:** segmented button
**Staging Review ticket:** [Staging Review: Design System - Segmented Button Component](https://github.com/department-of-veterans-affairs/va.gov-team/issues/111385)

## Findings details
**VA.gov Experience Standard - issue:** [User can safely use VA.gov or VA Mobile App products](https://design.va.gov/about/experience-standards/#user-can-safely-use-vagov-products)
**VA.gov Experience Standard - category:** [Health and Safety](https://design.va.gov/about/experience-standards/#category-health-and-safety)
**This is an issue with the:** code
**High-priority:** No
**Collab Cycle Reviewer:** @rsmithadhoc (Accessibility)

### Description

When clicking a different option, the focus indicator changes to the new selection, but there is a delay on changing the white background of the initial selection. This causes a flicker effect. I did not experience this when using the keyboard, only mouse.

This is not enough to fail the WCAG standard of three flashes per second or below, but could still affect users with photosensitive conditions.

<details><summary>Link, screenshot or steps to recreate</summary>

![Image](https://github.com/user-attachments/assets/54758f0f-32e4-4160-9b42-08c2b03af90d)

</details>

## Recommended action
Troubleshoot and fix the delay when clicking another option.

## References

- **Accessibility Defect Severity:** [4: Low. Consider fixing or exploring in 2-4 sprints post-launch.](https://depo-platform-documentation.scrollhelp.site/developer-docs/accessibility-defect-severity-rubric#Accessibilitydefectseverityrubric-a11y-defect-4)
- **WCAG Success Criteria:** [WCAG 2.3.1 Three Flashes or Below Threshold - Level A](https://www.w3.org/TR/WCAG22/#three-flashes-or-below-threshold)
- **Modality:** N/A
- **Design System Pattern or Template:** N/A
- **Design System Foundation:** N/A
- **Content Style Guide:** N/A
- **Context:** N/A

---

### Next Steps for DST

_How did this touchpoint go? [Give feedback on the Collaboration Cycle](https://dj540s05.optimalworkshop.com/questions/f54355271969f160ef41182ba9ceb61c) at any time._
- Close the ticket when the issue has been resolved or validated by your Product Owner
- If your team has additional questions or needs Governance help validating the issue, please comment on the ticket
- If this ticket has a `high-priority` label, please address as soon as possible so VFS teams are not impacted
- If this ticket has a `consider` label, please consider for future implementation

**Comment by powellkerry** [comment]
@amyleadem it looks like adding a mouseDown event listener, for earlier execution, to update the selected will resolve this. It seems a bit redundant though

```
                onMouseDown={() => {
                  this.selected = index;
                }}
                onClick={() => {
                  this.selected = index;
                  this.handleClick(buttonItem);
                }}
```

**Comment by RyanMunsch** [comment]
@amyleadem and @powellkerry, I have a branch with a similar (but more involved) solution to Kerry's above.

I added a state value and corresponding class to track whether the mouse is down and then wrapped some of the CSS selectors with a selector for that new class.

Both of these feel a little not ideal to me, but updating `this.selected` multiple times felt like it could lead to some race conditions or further confusion. We could probably get away with that and make sure to add a comment explaining why it's necessary.

```JSX
const containerClass = classnames({
  'usa-button-group': true,
  'va-segmented-button': true,
  'mouse-down': this.mouseDown,
});

[...]

onClick={() => {
  this.selected = index;
  this.handleClick(buttonItem);
}}
onMouseDown={() => this.mouseDown = true}
onMouseUp={() => this.mouseDown = false}
```

```CSS
/*
  Wrap active/focus styles in a "mouse-down" class to prevent the flicker effect
  when a new button is pressed while another is active.
*/
.va-segmented-button:not(.mouse-down) {
  .va-segmented-button__button:active,
  .va-segmented-button__button[aria-pressed=true] {
    background-color: var(--vads-color-white);
    box-shadow: 0 4px 0 0 var(--vads-color-white);
    position: relative;

    &::after {
      content: attr(data-label) / "";
      background: var(--vads-color-white);
      border-radius: 4px;
      font-weight: 700;
      position: absolute;
      top: 4px;
      left: 0;
      width: 100%;
      height: calc(100% - 4px);
      display: flex;
      align-items: center;
      justify-content: center;
    }
  }

  .va-segmented-button__button[aria-pressed=true]:focus,
  .va-segmented-button__button:active:focus,
  .va-segmented-button__button[aria-pressed="true"]:active:focus {
    background-color: var(--vads-color-action-focus-on-light);
    box-shadow: 0 4px 0 0 var(--vads-color-base);

    &::after {
      background-color: var(--vads-color-action-focus-on-light);
    }
  }
}
```

**Comment by babsdenney** [comment]
We had a meeting yesterday to discuss some design options for addressing this issue. Here are some of the options to consider, and there may be others that haven't been thought of yet. 

- **Option 1:** Do nothing. This is an edge case that only becomes noticeable when quickly switching through the buttons, so it may not be obvious when using it normally. We also checked the [UKDS Tabs](https://design-system.service.gov.uk/components/tabs/default/#past-year), and they have the same issue, but it isn't as clear due to the other elements surrounding the tabs. This probably isn't the best option, as we want to solve for all edge cases. However, you can see that the issue becomes less obvious when there is more content on the page. 
- **Option 2:** Add a transition. In the example @t-michaud created, you can see that the active state has a fade transition to reduce the appearance of a flicker.
- **Option 3:** Focus visible. Only showing the focus states while tabbing would solve the issue, but this is not the most recommended path, as it creates an inconsistent experience for the components.
- **Option 4:** Explore alternative visual designs for the Active (White button) state. The flicker is most noticeable because of the white active state. There may be other design options we could consider.

@davidakennedy mentioned - 

> One thing I think may be true: How the combination of :focus and :active states are implemented across browsers may be different. Something to be mindful of when thinking about a solution.

**Comment by caw310** [comment]
The team met again to discuss this issue and decided that we  are going to revert to the old focus style. 
We are seeing similar edge case behavior with other new components we are working on (ie. Sidenav and new Tabs). Rather than creating one-offs for multiple components we will do [Discovery: Accessibility Review of the new focus styles](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/4467#top) so that we can come up with a way to address all these quirks. 

@shiragoodman Can we close this issue?
